algorithms/src/snark/marlin/ahp/prover/state.rs

```diff
@@ -70,95 +62,164 @@ pub struct State<'a, F: PrimeField, MM: MarlinMode> {
 
     /// A list of polynomials corresponding to the interpolation of the public input.
     /// The length of this list must be equal to the batch size.
-    pub(super) x_poly: Vec<DensePolynomial<F>>,
-
-    /// The first round oracles sent by the prover.
-    /// The length of this list must be equal to the batch size.
-    pub(in crate::snark) first_round_oracles: Option<Arc<super::FirstOracles<F>>>,
+    pub(super) x_polys: Vec<DensePolynomial<F>>,
 
     /// Randomizers for z_b.
     /// The length of this list must be equal to the batch size.
     pub(super) mz_poly_randomizer: Option<Vec<F>>,
 
-    /// The challenges sent by the verifier in the first round
-    pub(super) verifier_first_message: Option<verifier::FirstMessage<F>>,
-
     /// Polynomials involved in the holographic sumcheck.
     pub(super) lhs_polynomials: Option<[DensePolynomial<F>; 3]>,
-    /// Polynomials involved in the holographic sumcheck.
-    pub(super) sums: Option<[F; 3]>,
 }
 
+/// State for the AHP prover.
+pub struct State<'a, F: PrimeField, MM: MarlinMode> {
+    pub(super) circuit_specific_states: BTreeMap<&'a Circuit<F, MM>, CircuitSpecificState<F>>,
+    pub(super) total_instances: usize,
+    pub(super) max_num_constraints: usize,
+    /// The challenges sent by the verifier in the first round
+    pub(super) verifier_first_message: Option<verifier::FirstMessage<'a, F>>,
+    /// The first round oracles sent by the prover.
+    /// The length of this list must be equal to the batch size.
+    pub(in crate::snark) first_round_oracles: Option<Arc<super::FirstOracles<'a, F, MM>>>,
+    pub(in crate::snark) max_non_zero_domain: EvaluationDomain<F>,
+    pub(in crate::snark) max_constraint_domain: EvaluationDomain<F>,
+}
+
+pub type PaddedPubInputs<F> = Vec<F>;
+pub type PrivateInputs<F> = Vec<F>;
+pub type Za<F> = Vec<F>;
+pub type Zb<F> = Vec<F>;
+pub struct Assignments<F>(
+    pub PaddedPubInputs<F>,
+    pub PrivateInputs<F>,
+    pub Za<F>,
+    pub Zb<F>
+);
```

Comment:
Let's make things `pub(crate)`/`pub(super)` where appropriate, and add some doc-comments.

Reply:
Done for any new types and structures this PR adds